core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1520,29 +1522,66 @@ class KafkaApis(val requestChannel: RequestChannel,
-          .setThrottleTimeMs(requestThrottleMs)
-          .setErrorCode(error.code()))
-        trace("Sending leave group response %s for correlation id %d to client %s."
-          .format(response, request.header.correlationId, request.header.clientId))
-        response
-      }
-      sendResponseMaybeThrottle(request, createResponse)
+    // Before version 3, leave group request is still in single mode
+    val members = if (leaveGroupRequest.version <= 2) {
+      List(new MemberIdentity()
+        .setMemberId(leaveGroupRequest.data.memberId))
+    } else {
+      leaveGroupRequest.data.members.asScala.toList
     }
 
-    if (!authorize(request.session, Read, Resource(Group, leaveGroupRequest.data().groupId(), LITERAL))) {
-      sendResponseMaybeThrottle(request, requestThrottleMs =>
+    if (!authorize(request.session, Read, Resource(Group, leaveGroupRequest.data.groupId, LITERAL))) {
+      sendResponseMaybeThrottle(request, requestThrottleMs => {
         new LeaveGroupResponse(new LeaveGroupResponseData()
           .setThrottleTimeMs(requestThrottleMs)
-          .setErrorCode(Errors.GROUP_AUTHORIZATION_FAILED.code())))
+          .setErrorCode(Errors.GROUP_AUTHORIZATION_FAILED.code)
+          .setMembers(
+            members.map (
+              member => new MemberResponse()
+                .setMemberId(member.memberId)
+                .setGroupInstanceId(member.groupInstanceId)
+                .setErrorCode(Errors.GROUP_AUTHORIZATION_FAILED.code)).asJava
+          )
+        )
+      })
     } else {
-      // let the coordinator to handle leave-group
+      def sendResponseCallback(leaveGroupResults: List[LeaveGroupResult]) {
+        val memberResponses = leaveGroupResults.map(
+          leaveGroupResult =>
+            new MemberResponse()
+              .setErrorCode(leaveGroupResult.error.code)
+              .setMemberId(leaveGroupResult.memberId)
+              .setGroupInstanceId(leaveGroupResult.groupInstanceId.orNull)
+        )
+        def createResponse(requestThrottleMs: Int): AbstractResponse = {
+          val batchLeaveGroupResponse = if (leaveGroupRequest.version <= 2) {
+            if (leaveGroupResults.size != 1) {
+              throw new IllegalStateException(s"Singleton leave group request shouldn't have more than 1 " +
+                s"response, while actually get ${leaveGroupResults.size}")
+            }
+            new LeaveGroupResponse(
+              new LeaveGroupResponseData()
+                .setThrottleTimeMs(requestThrottleMs)
+                .setErrorCode(memberResponses.head.errorCode)
+            )
+          } else {
+            new LeaveGroupResponse(
+              new LeaveGroupResponseData()
+                .setThrottleTimeMs(requestThrottleMs)
+                .setMembers(memberResponses.asJava)
+            )
+          }
+
+          trace("Sending leave group response %s for correlation id %d to client %s."
```

Comment:
We may be able to remove this. I think it would be covered by request logs.

Reply:
Which request log you mean specifically?

Reply:
Kafka has a request log which can already be used to log incoming requests and outgoing responses. So I think this message is redundant and we can remove it.

Reply:
Sounds good